proposals/0000-amending-mrp.rst

```diff
@@ -0,0 +1,256 @@
+Other links of interest include the `current proposal <https://gitlab.haskell.org/ghc/ghc/-/wikis/proposal/monad-of-no-return>`_,
+the `original proposal <https://mail.haskell.org/pipermail/libraries/2015-September/026121.html>`_,
+rhendric's `current implementation <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3174>`_
+of the amended phase 2 (see below) and associated `issue <https://gitlab.haskell.org/ghc/ghc/-/issues/25783>`_.
+There is also the enablement of the warnings by default on `gitlab <https://gitlab.haskell.org/ghc/ghc/-/merge_requests/3174>`_
+and its `own proposal <https://github.com/ghc-proposals/ghc-proposals/pull/314>`_.
+
+
+Proposed Change Specification
+-----------------------------
+To summarise the proposal to date:
+
+* When making ``Applicative`` a superclass of Monad, there are now redundant 
+  methods in ``Monad`` in ``return`` and ``(>>)``, as these can always (modulo
+  performance) be replaced with ``pure`` and ``(*>)`` respectively
+* The first phase was completed in 2015, that of adding a warning (off by 
+  default) that warns when definitions of ``return`` and ``(>>)`` are present
+  that aren't just making them equivalent to ``pure`` and ``(*>)`` respectively.
+  This is ``-Wnoncanonical-monad-instances``
+* In ~2020 this warning was enabled by default in 9.2.1
+
+The next phases of the proposal have not been moved forward until very recently.
+
+* Phase 2 requires that the previous warning is now an error and that 
+  ``return`` and ``(>>)`` are top level bindings, but also that canonical
+  definitions are ignored
+* Phase 3 would then start warning about canonical ``return`` and ``(>>)``
+  definitions
+* Phase 4 would remove support for any sort of override, making that an error
+
+A lawful method is simply a method that follows the laws of the typeclass.
+A canonical method is one that is always definitionally the same, and has no 
+greater power. For example, ``return`` and ``pure`` must always be equivalent, 
+and there is no reason that they should not be defined in terms of each other. 
+The proposal seeks to eliminate the presence of the excess canonical methods
+that can be defined in ``Monad``.
+
+This proposal would adjust the next phases to reduce the number of changes
+neccessary in GHC as follows:
+
+* Phase 2 makes ``-Wnoncanonical-monad-instances`` a compiler error by default,
+  and adds a ``-Wredundant-canonical-monad-method`` warning that makes canonical
+  method definitions (only ``return = pure``) warn
+* Phase 3 would then move ``return`` and ``(>>)`` to the top level, and remove
+  the original and new warnings (after removing these methods from ``Monad`` the
+  compiler will emit an error like for any other extraneous typeclass instance
+  method definition)
+
+  * Phase 3 shouldn't be  (fully) implemented until research is done as to what 
+    we want the eventual definition of ``(>>)`` to be
```

Comment:
Arguably we should figure this out before proceeding with phase 2. The one thing that doesn't seem to be controversial about this part of the proposal is that `(*>)` and `(>>)` should always be identical. That means that either the current default implementation of `(>>)` should become the default implementation of `(*>)` (`Monad` constraint and all, requiring any non-`Monoid` `Applicative` instances to hand-write a `(*>)` definition), or the default implementation of `(*>)` remains and we use some amount of compiler intelligence to push users toward writing `a *> b = a >>= \_ -> b` when appropriate. The choice here affects the suggestions offered by the compiler when we nudge users towards canonicalizing their instances.